src/Snak/SnakList.php

```diff
@@ -15,17 +19,247 @@
  * @author Jeroen De Dauw < jeroendedauw@gmail.com >
  * @author Addshore
  */
-class SnakList extends HashArray {
+class SnakList extends ArrayObject implements Hashable, Comparable {
 
 	/**
-	 * @see GenericArrayObject::getObjectType
+	 * Maps snak hashes to their offsets.
+	 *
+	 * @var array [ snak hash (string) => array [ snak offset (string|int) ] | snak offset (string|int) ]
+	 */
+	private $offsetHashes = [];
+
+	/**
+	 * @var integer
+	 */
+	private $indexOffset = 0;
+
+	/**
+	 * @see ArrayObject::__construct
+	 *
+	 * @param array|Traversable|null $input
+	 * @param int $flags
+	 * @param string $iteratorClass
+	 *
+	 * @throws InvalidArgumentException
+	 */
+	public function __construct( $input = null, $flags = 0, $iteratorClass = 'ArrayIterator' ) {
+		parent::__construct( [], $flags, $iteratorClass );
+
+		if ( $input !== null ) {
+			if ( !is_array( $input ) && !( $input instanceof Traversable ) ) {
+				throw new InvalidArgumentException( '$input must be an array or Traversable' );
+			}
+
+			foreach ( $input as $offset => $value ) {
+				$this->offsetSet( $offset, $value );
+			}
+		}
+	}
+
+	/**
+	 * Finds a new offset for when appending an element.
+	 * The base class does this, so it would be better to integrate,
+	 * but there does not appear to be any way to do this...
+	 *
+	 * @return integer
+	 */
+	private function getNewOffset() {
+		while ( $this->offsetExists( $this->indexOffset ) ) {
+			$this->indexOffset++;
+		}
+
+		return $this->indexOffset;
+	}
+
+	/**
+	 * Gets called before a new element is added to the list.
+	 *
+	 * Should return a boolean. When false is returned the element
+	 * does not get added to the ArrayObject.
+	 *
+	 * @since 0.1
+	 *
+	 * @param int|string $index
+	 * @param Hashable $hashable
+	 *
+	 * @return bool
+	 */
+	private function preSetElement( $index, $hashable ) {
+		$hash = $hashable->getHash();
+
+		$hasHash = $this->hasSnakHash( $hash );
+
+		if ( $hasHash ) {
+			return false;
+		}
+
+		if ( $hasHash ) {
```

Comment:
This is now dead code. See #712, where I remove this entire feature from the base class. As before I will happily rebase this patch when #712 is merged.